### ISSUE TITLE
OpenTelemetry support

### ISSUE THREAD

**Issue by donbeave** [jOOQ/jOOQ#11965]
Hello, @lukaseder ! 

First of all I want to say thank you for creating jOOQ. I love it a lot and I think this is a game-changer library in Java world. For many apps it's something like you can see on Instagram on advertising photos, like the fat person was before and the slim person was after. The jOOQ is that result of slim pretty app. Sorry for the offtopic. 😀

With using jooq in all my microservices and last initiative with adding one standard for all tracing data I realized that jooq official library does not supports opentelemetry or opentelemetery-java-instrumentation does not support jooq out of box. Maybe it's time to fill that gap. Currently, I'm working on OpenTelemetry library for the Micronaut framework and I can help (if needed) with making support for jOOQ framework. If you think this code (jooq opentelemetry module) can be part of official jooq repository I can make a PR.

**Comment by lukaseder** [comment]
Thanks for your suggestion, @donbeave. I can't say I get the point of making an inappropriate body shaming analogy in this context.

There are many telementry APIs out there. E.g. JFR comes to mind. I haven't heard of opentelemetry yet. It doesn't seem to have a lot of traction yet: https://stackoverflow.com/questions/tagged/open-telemetry

It does sound like an interesting topic for third parties to integrate. The best place to start looking for possible integrations is the `ExecuteListener` SPI, which is very versatile for many similar such purposes: https://www.jooq.org/doc/latest/manual/sql-execution/execute-listeners

I hope this helps. I currently don't see this as part of jOOQ.

**Comment by donbeave** [comment]
> Thanks for your suggestion, @donbeave. I can't say I get the point of making an inappropriate body shaming analogy in this context.

Sorry for this. Actually, it's a bad comparison, just want to say the same application after the switch to the jOOQ became more lightweight compare to other solutions.

>  I haven't heard of opentelemetry yet.

Hmm, probably you heard of [OpenTracing](https://opentracing.io) or [OpenCensus](https://opencensus.io) before. Both of these projects have merged to form [OpenTelemetry](https://opentelemetry.io), a member of Cloud Native Computing Foundation.

The adoption for it is growing, not fast, but quite stable with support from top tech companies:

https://newrelic.com/blog/best-practices/opentelemetry-opentracing-opencensus
https://cloud.google.com/blog/products/operations/opentelemetry-specification-enables-standardized-tracing
https://github.blog/2021-05-26-why-and-how-github-is-adopting-opentelemetry/
https://www.dynatrace.com/news/press-release/dynatrace-teams-with-google-and-microsoft-on-opentelemetry/

**Comment by lukaseder** [comment]
> Hmm, probably you heard of [OpenTracing](https://opentracing.io) or [OpenCensus](https://opencensus.io) before

I have not :)

> Both of these projects have merged to form [OpenTelemetry](https://opentelemetry.io), a member of Cloud Native Computing Foundation.
> 
> The adoption for it is growing, not fast, but quite stable with support from top tech companies

That's great, and certainly good to keep an eye out for what they're doing.

But again, it's super easy to connect jOOQ's `ExecuteListener` with whatever telemetry or logging APIs to get tailor-made events, so I definitely don't want to rush anything in the near future. jOOQ has been around for 12 years. I've seen "standards" come and go all the time, and I never regretted sitting them out. Take reactive programming, for example. It took ages for it to produce a "standard" SPI like R2DBC (quotes around "standard", because it still needs to be proven. It has much potential, but it is not yet widely adopted by vendors). I finally decided to integrate with this most promising take at reactive SQL so far: https://github.com/jOOQ/jOOQ/issues/11700. It's been a ton of work, and a ton of risk, because

1. The SPI is a moving target, not stable yet. I've been involved a few times with the design of some edge cases. This is just to stress again how much work such an integration causes. But here, jOOQ will be a first mover.
2. There's Loom by Oracle, which might just as well obsolete 80% of all reactive-stream based API use-cases, because it promises to implement both push and pull models uniformly behind the scenes without changing the sequental, imperative-style programming model, which is *much* easier to understand than the reactive streams alternative.
3. There were competing attempts at standardising reactive SQL before (e.g. ADBA), and many third party drivers that didn't implement any standard SPI. Which horse to bet on? It was impossible to tell, though I'm glad that Oracle dropped ADBA.

R2DBC is very strategic to the core of jOOQ's value proposition, so the investment is justified. Telemetry? Much less so. jOOQ definitely doesn't have to be a first mover in this area. A typical jOOQ based application would much more likely find most issues of that kind in a tool chain like Oracle Enterprise Manager / Cloud Control, and classic ASH / AWR reports (or some IBM / Microsoft / Amazon equivalent), and there's nothing jOOQ needs to do to integrate with those, they work out of the box. Additionally, tools like Dynatrace, App Dynamics, and others support collecting data directly based on JDBC API calls, which is awesome! JDBC is such a great API for such purposes, and so easy to instrument / proxy.

In summary, I'd need to see:

- More market adoption of "Open Telemetry"
- More end user use-cases than marketing blog posts
- More users requesting precisely this
- More comparable integrations by other comparable APIs (and success stories from them)
- Comparisons of telemetry APIs, e.g. this one with Java Flight Recorder, and many others, and strong arguments in favour of one over the other, or a path to implement support for them all.

From experience, I'd say, this is a topic worth re-visiting in 3-5 years.

**Comment by donbeave** [comment]
Just in case if someone finds this issue in Google or somewhere else, I have implemented JDBC instrumentation for `opentelemetry-java-instrumentation` library, please check this PR: https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/3367.
